File: dev/vm_jvm/lmx/LmxOccurrence.java
package lmx;

import java.util.IdentityHashMap;
import java.util.Map;

/**
 * Isolated JVM ABI for an Lmx occurrence {@code {node, len, data}}.
 *
 * <p>{@code node} is the physical reference to the lexical parent occurrence,
 * or {@code null} for an independent root ({@code node = 0} in the native model).
 * It is not a kind/type tag (see docs/LMX_semantics.en.md ?3, ?8, ?9).
 *
 * <p>{@code merge} copies the used mutable graph into a new root with one
 * source-to-copy identity map: shared mutable targets stay shared, {@code node}
 * links are rewritten into the copy, operands are unchanged. Admitted terminals
 * ({@link LmxTerminal}) and non-occurrence leaves keep identity
 * (docs/LMX_semantics.en.md ?19, ?22).
 */
public final class LmxOccurrence {
    /** Lexical parent occurrence, or {@code null} if independent. */
    private final LmxOccurrence node;
    private final Object[] data;

    private LmxOccurrence(LmxOccurrence node, Object[] data) {
        if (data == null) {
            throw new IllegalArgumentException("data");
        }
        this.node = node;
        this.data = data;
    }

    /** Independent root: {@code node == null}. */
    public static LmxOccurrence independent(Object... fields) {
        LmxOccurrence root = new LmxOccurrence(null, copyFields(fields));
        bindNestedParents(root);
        return root;
    }

    /** Nested occurrence whose lexical parent is exactly {@code parent}. */
    public static LmxOccurrence nested(LmxOccurrence parent, Object... fields) {
        if (parent == null) {
            throw new IllegalArgumentException("parent");
        }
        LmxOccurrence o = new LmxOccurrence(parent, copyFields(fields));
        bindNestedParents(o);
        return o;
    }

    public LmxOccurrence node() {
        return node;
    }

    public int len() {
        return data.length;
    }

    public Object child(int index) {
        checkBounds(index);
        return data[index];
    }

    /** Assignment writes {@code data[index]}, never {@code len}. */
    public void setChild(int index, Object value) {
        checkBounds(index);
        data[index] = value;
    }

    /**
     * Graph-copy merge (?19): fresh independent root; direct fields = operand
     * fields in order; mutable {@link LmxOccurrence} graph copied via identity map;
     * {@link LmxTerminal} and other non-occurrence leaves shared by identity.
     * New root {@code node} is {@code null} (isolated expression / zero parent).
     */
    public static LmxOccurrence merge(LmxOccurrence... parts) {
        if (parts == null || parts.length == 0) {
            throw new IllegalArgumentException("parts");
        }
        int n = 0;
        for (LmxOccurrence p : parts) {
            if (p == null) {
                throw new IllegalArgumentException("null part");
            }
            n += p.len();
        }
        Map<LmxOccurrence, LmxOccurrence> map = new IdentityHashMap<LmxOccurrence, LmxOccurrence>();
        Object[] fields = new Object[n];
        LmxOccurrence root = new LmxOccurrence(null, fields);
        int k = 0;
        for (LmxOccurrence p : parts) {
            for (int i = 0; i < p.len(); i++) {
                fields[k++] = copyRef(p.child(i), map, root);
            }
        }
        return root;
    }

    private static Object copyRef(Object value, Map<LmxOccurrence, LmxOccurrence> map, LmxOccurrence container) {
        if (value instanceof LmxOccurrence) {
            return copyOccurrence((LmxOccurrence) value, map, container);
        }
        // Leaves and admitted terminals: share identity (primitives, LmxTerminal, ?).
        return value;
    }

    private static LmxOccurrence copyOccurrence(
            LmxOccurrence src, Map<LmxOccurrence, LmxOccurrence> map, LmxOccurrence parentForCopy) {
        LmxOccurrence existing = map.get(src);
        if (existing != null) {
            return existing;
        }
        Object[] data = new Object[src.len()];
        LmxOccurrence dst = new LmxOccurrence(parentForCopy, data);
        map.put(src, dst);
        for (int i = 0; i < src.len(); i++) {
            data[i] = copyRef(src.child(i), map, dst);
        }
        return dst;
    }

    /** Nested mutable children in {@code data} get {@code node == container}. */
    private static void bindNestedParents(LmxOccurrence container) {
        Map<LmxOccurrence, LmxOccurrence> rebound = new IdentityHashMap<LmxOccurrence, LmxOccurrence>();
        for (int i = 0; i < container.data.length; i++) {
            Object c = container.data[i];
            if (c instanceof LmxOccurrence) {
                LmxOccurrence child = (LmxOccurrence) c;
                if (child.node == container) {
                    continue;
                }
                LmxOccurrence fixed = rebound.get(child);
                if (fixed == null) {
                    fixed = reparent(child, container);
                    rebound.put(child, fixed);
                }
                container.data[i] = fixed;
            }
        }
    }

    private static LmxOccurrence reparent(LmxOccurrence src, LmxOccurrence newParent) {
        Object[] data = new Object[src.len()];
        System.arraycopy(src.data, 0, data, 0, src.len());
        LmxOccurrence o = new LmxOccurrence(newParent, data);
        bindNestedParents(o);
        return o;
    }

    private static Object[] copyFields(Object[] fields) {
        Object[] copy = new Object[fields.length];
        System.arraycopy(fields, 0, copy, 0, fields.length);
        return copy;
    }

    private void checkBounds(int index) {
        if (index < 0 || index >= data.length) {
            throw new IndexOutOfBoundsException(
                    "LmxOccurrence index " + index + " not in [0," + data.length + ")");
        }
    }
}

File: dev/vm_cil/Lmx/LmxOccurrence.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Lmx
{
    /// <summary>Isolated CIL ABI for {node, len, data}. Store does not reparent.</summary>
    public sealed class LmxOccurrence
    {
        private LmxOccurrence node;
        private readonly object[] data;

        private LmxOccurrence(LmxOccurrence node, object[] data)
        {
            if (data == null) throw new ArgumentException("data");
            this.node = node;
            this.data = data;
        }

        private LmxOccurrence(int len)
        {
            this.node = null;
            this.data = new object[len];
        }

        public static LmxOccurrence Independent(params object[] fields)
        {
            return new LmxOccurrence(null, CopyFields(fields));
        }

        public static LmxOccurrence Nested(LmxOccurrence parent, params object[] fields)
        {
            if (parent == null) throw new ArgumentException("parent");
            return new LmxOccurrence(parent, CopyFields(fields));
        }

        public LmxOccurrence Node { get { return node; } }
        public int Len { get { return data.Length; } }

        public object Child(int index)
        {
            CheckBounds(index);
            return data[index];
        }

        public void SetChild(int index, object value)
        {
            CheckBounds(index);
            data[index] = value;
        }

        public static LmxOccurrence Merge(params LmxOccurrence[] parts)
        {
            if (parts == null || parts.Length == 0) throw new ArgumentException("parts");
            var map = new Dictionary<LmxOccurrence, LmxOccurrence>(ReferenceEqualityComparer.Instance);
            int n = 0;
            for (int p = 0; p < parts.Length; p++)
            {
                if (parts[p] == null) throw new ArgumentException("null part");
                n += parts[p].Len;
                for (int i = 0; i < parts[p].Len; i++)
                {
                    object c = parts[p].Child(i);
                    if (c is LmxOccurrence) Ensure((LmxOccurrence)c, map);
                }
            }
            FinishCopy(map);
            object[] fields = new object[n];
            LmxOccurrence root = new LmxOccurrence(null, fields);
            int k = 0;
            for (int p = 0; p < parts.Length; p++)
            {
                for (int i = 0; i < parts[p].Len; i++)
                    fields[k++] = MapRef(parts[p].Child(i), map);
            }
            return root;
        }

        private static void Ensure(LmxOccurrence src, Dictionary<LmxOccurrence, LmxOccurrence> map)
        {
            if (src == null || map.ContainsKey(src)) return;
            map[src] = new LmxOccurrence(src.Len);
            Ensure(src.node, map);
            for (int i = 0; i < src.Len; i++)
            {
                object c = src.Child(i);
                if (c is LmxOccurrence) Ensure((LmxOccurrence)c, map);
            }
        }

        private static void FinishCopy(Dictionary<LmxOccurrence, LmxOccurrence> map)
        {
            foreach (var e in map)
            {
                LmxOccurrence src = e.Key;
                LmxOccurrence dst = e.Value;
                if (src.node == null) dst.node = null;
                else
                {
                    LmxOccurrence mapped;
                    if (!map.TryGetValue(src.node, out mapped) || mapped == null)
                        throw new InvalidOperationException("ancestor missing from copy map");
                    dst.node = mapped;
                }
                for (int i = 0; i < src.Len; i++)
                    dst.data[i] = MapRef(src.Child(i), map);
            }
        }

        private static object MapRef(object value, Dictionary<LmxOccurrence, LmxOccurrence> map)
        {
            LmxOccurrence occ = value as LmxOccurrence;
            if (occ != null)
            {
                LmxOccurrence mapped;
                if (!map.TryGetValue(occ, out mapped) || mapped == null)
                    throw new InvalidOperationException("occurrence missing from copy map");
                return mapped;
            }
            return value;
        }

        private static object[] CopyFields(object[] fields)
        {
            if (fields == null) return new object[0];
            object[] copy = new object[fields.Length];
            Array.Copy(fields, copy, fields.Length);
            return copy;
        }

        private void CheckBounds(int index)
        {
            if (index < 0 || index >= data.Length)
                throw new IndexOutOfRangeException(
                    "LmxOccurrence index " + index + " not in [0," + data.Length + ")");
        }

        private sealed class ReferenceEqualityComparer : IEqualityComparer<LmxOccurrence>
        {
            public static readonly ReferenceEqualityComparer Instance = new ReferenceEqualityComparer();
            public bool Equals(LmxOccurrence x, LmxOccurrence y) { return object.ReferenceEquals(x, y); }
            public int GetHashCode(LmxOccurrence obj) { return RuntimeHelpers.GetHashCode(obj); }
        }
    }
}
